src/librustc_trans/back/link.rs

```diff
@@ -237,7 +237,7 @@ pub fn default_output_for_target(sess: &Session) -> config::CrateType {
 /// Checks if target supports crate_type as output
 pub fn invalid_output_for_target(sess: &Session,
                                  crate_type: config::CrateType) -> bool {
-    match (sess.target.target.options.dynamic_linking,
+    match (sess.target.target.options.dynamic_linking && !sess.target_is_crt_static(),
```

Comment:
Following that logic, it should also be possible to build MUSL dylibs (`.so` libraries) that statically link to libc but link dynamically to other libraries I think (maybe not?). Do we want to support that as well?

If not then I can just add a `!musl ||` here I think.